### PR TITLE
Open in Excel

### LINKED ISSUE #19: Export to Excel
When a table is processed so all the table data is visible it should be nice to have an option to export this data to excel.

### PR DESCRIPTION
**Pull request by Busschers**
Added Excel Export

### CODE REVIEW

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,73 @@
+codeunit 50138 "OVW Export to Excel"
+{
+    trigger OnRun()
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: You can remove the OnRun trigger if it's empty

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    begin
+
+    end;
+
+    procedure Rec2Excel(TableID: Integer)
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: A few things here:

(1) Make this procedure internal, unless there is a need to access it from outside the app
(2) Procedures always start with a verb and avoid using the 2 as a substitute for "to". Consider renaming this to ExportRecordToExcel.

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    var
+        RecRef: RecordRef;
+    begin
+        RecRef.Open(TableID);
+        if not RecRef.ReadPermission then
+            exit;
+
+        RowNo := 1;
+
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: Remove unnecessary whitespace

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            repeat
+                Ref2Excel(RecRef, TableID);
+            until RecRef.Next() = 0;
+        end;
+        TempExcelBuffer.CreateExcelBook(RecRef.Name); //change into table name?
+    end;
+
+    local procedure Ref2Excel(var Ref: RecordRef; TableID: Integer)
+    var
+        int: Integer;
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: Can you rename this to a standard naming convention? Something like FieldIndex or CurrentFieldIndex

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        if RecRef.FindSet() then begin
+            repeat
+                Ref2Excel(RecRef, TableID);
+            until RecRef.Next() = 0;
+        end;
+        TempExcelBuffer.CreateExcelBook(RecRef.Name); //change into table name?
+    end;
+
+    local procedure Ref2Excel(var Ref: RecordRef; TableID: Integer)
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: Same as before; I recommend ExportRecordRefToExcel

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    begin
+        ColumnNo := 1;
+        RowNo += 1;
+
+        for int := 1 to Ref.FieldCount() do begin
+            fRef2Excel(Ref, int, TableID);
+        end;
+    end;
+
+    local procedure fRef2Excel(var Ref: RecordRef; var i: Integer; TableID: Integer)
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: Same as before; I recommend ExportRecordRefToExcel

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            FRef.Class::Normal:
+                begin
+                    TempExcelBuffer.AddCell(RowNo, ColumnNo, FRef.Value, FRef.Caption, 1);
+                end;
+
+            FRef.Class::FlowField:
+                begin
+                    FRef.CalcField();
+                    TempExcelBuffer.AddCell(RowNo, ColumnNo, FRef.Value, FRef.Caption, 1);
+                end;
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al:53]: There are more field classes than the two in the case loop. You will need to handle those or add and "else". If you want this extensible, you might want to add an integration event publisher on the _else_ case.

File: SRC/Pages/OVWObjectViewer.Page.al [hunk 8]
```diff
@@ -82,6 +82,22 @@ page 50130 "OVW Object Viewer"
...
             }
+
+            action(ExportToExcel)
+            {
+                ApplicationArea = All;
+                Caption = 'Export To Excel';
+                Image = Excel;
+                Promoted = true;
+                PromotedCategory = Process;
+                PromotedIsBig = true;
```
Comment (Peter-Conijn) [inline SRC/Pages/OVWObjectViewer.Page.al]: PromotedIsBig is an obsolete option and can be left out.

Also: please add a tooltip to your action.

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        field(50131; "OVW Value is BLOB"; Boolean)
+        {
+            DataClassification = SystemMetadata;
+        }
+    }
+
+    procedure AddCell(pintRow: Integer; var pintColumn: Integer; pvarValue: Variant; ptxtHeading: Text[250]; pintColumnIncrement: Integer)
+    begin
+        AddCellWithFormatting(pintRow, pintColumn, pvarValue, ptxtHeading, false, false, false, 2, pintColumnIncrement); //34764+-
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al]: There is a documentation tag in this. Can you remove  that?

Would it also be possible to refactor your variables to use standard naming conventions instead of the Hungarian notation?

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        end;
+        if not "OVW Value is BLOB" then
+            "Cell Value as Text" := Format(CellValue);
+        if not Insert() then
+            Modify();
+        AddHeading(ColumnNo, CellHeading);
+        ColumnNo += ColumnIncrementValue;
+    end;
+
+    [Obsolete('This procedure is no longer used and will be deprecated. This warning will become an error in the future.', '19.1.20220408')]
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al]: Should this Obsolete tag be here? If so, remove the procedure and if not, please remove the tag.

And again, please refactor variable names to adhere to the Microsoft coding guidelines. For details please check the [AL Guidelines page.](https://alguidelines.dev/)

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    procedure AddFormula(pintRow: Integer; var pintColumn: Integer; ptxtFormula: Text[250]; ptxtHeading: Text[250]; pintColumnIncrement: Integer)
+    begin
+        Init();
+        Validate("Row No.", pintRow);
+        Validate("Column No.", pintColumn);
+        "Cell Value as Text" := '';
+        Formula := ptxtFormula;
+        if not Insert() then
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al]: Please do not use this method; this can cause severe drains on performance.

Use an "if not Get() then Insert()" pattern instead.

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+    procedure CreateFile(SheetName: Text)
+    begin
+        CreateNewBook(SheetName);
+        WriteSheet(CopyStr(SheetName, 1, 80), CompanyName, UserId);
+        CloseBook();
+        OpenExcel();
+    end;
+
+    [Obsolete('This procedure is no longer used and will be deprecated. This warning will become an error in the future.', '19.1.20220408')]
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al]: Please remove this Obsolete tag.

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    var
+        WriteStream: OutStream;
+    begin
+        TempBlob.CreateOutStream(WriteStream, TextEncoding::UTF8);
+        WriteStream.WriteText(Input);
+    end;
+
+    var
+        TempBlob: Codeunit "Temp Blob";
+
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al:132]: Unnecessary whitespace

File: SRC/TableExtensions/OVWExcelBuffer.TableExtension.al [hunk 14]
```diff
@@ -0,0 +1,164 @@
+tableextension 50131 "OVW Excel Buffer" extends "Excel Buffer"
+{
+    fields
+    {
+        field(50131; "OVW Value is BLOB"; Boolean)
```
Comment (Peter-Conijn) [inline SRC/TableExtensions/OVWExcelBuffer.TableExtension.al:5]: Please add a caption to this field

File: SRC/Pages/OVWObjectViewer.Page.al [hunk 15]
```diff
@@ -194,4 +210,15 @@ page 50130 "OVW Object Viewer"
             exit;
         exit(OVWRecordCountMgt.GetRecordCount(ForObjectID));
     end;
+
+    local procedure DoExportToExcel()
```
Comment (Peter-Conijn) [inline SRC/Pages/OVWObjectViewer.Page.al:215]: This is a good data validation, but you might also add some code on the action in the page extension so that it is only enabled when the object type is Table.

File: SRC/Codeunits/OVWExportToExcel.Codeunit.al [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    trigger OnRun()
+    begin
+
+    end;
+
+    procedure Rec2Excel(TableID: Integer)
+    var
+        RecRef: RecordRef;
+    begin
+        RecRef.Open(TableID);
```
Comment (Peter-Conijn) [inline SRC/Codeunits/OVWExportToExcel.Codeunit.al]: This might be improved by refactoring this in a [Try Function](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-handling-errors-using-try-methods).

**Review by Peter-Conijn — CHANGES_REQUESTED** [review]
This is a useful functionality, but there are some things that can be approved. I have added some comments in the code.

**Comment by Busschers** [comment]
Added a new commit with changes based on your comments.

**Review by Peter-Conijn — APPROVED** [review]

**Comment by Peter-Conijn** [comment]
Looks good. Merge approved
